### DISCUSSION TITLE
Arbitrary channel mapping in the receiver for CRSF output

### DISCUSSION THREAD

**Discussion by robhaswell** [ExpressLRS/ExpressLRS#1486] [Ideas]
As Matek has begun shipping CRSF to SBUS converters, it is possible to add ELRS to many existing flybarless units (flight controllers) which do not support CRSF . Unfortunately most of these units are proprietary and the channel order cannot be changed. Helicopters require 4 full-resolution channels (A,E,R and Collective) and an additional channel for Throttle for which 64 positions is sufficient. Typically the Collective is on CH6, and CH5 used to change flight modes.

To support flybarless units and other proprietary flight controllers it would be useful to be able to configure the receiver (possibly at compile-time) to map the channels received from the transmitter onto different channel numbers before outputting over CRSF.

Related: #1190

**Comment by robhaswell** [comment]
Hi, is it possible for this feature to be considered please?

**Reply by ru13ab** [reply]
Maybe this feature of the third version will help you.  https://github.com/ExpressLRS/ExpressLRS/pull/1572

**Reply by robhaswell** [reply]
V3 features look awesome, having 8 full-res channels is perfect. The one problem I can see is the requirement for 1-bit on AUX1 (CH5). Unfortunately CH5 tends to be used for flight mode switching which conflicts with the Arm semantics.

I notice in the V3 announcement that there is an image which demonstrates full resolution on CH5:

![image](https://camo.githubusercontent.com/174cb01c72ec6c9f5f8a1599a2de52ef39077deae27e9a191710141a8e11e07f/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3539363335303032323139313431353332302f3937363832393834333035343339353434322f333333487a2d31362e676966)

Is it possible somehow to get full resolution on CH5? Perhaps a custom fork which moves "Arm" onto a different channel?

**Reply by ru13ab** [reply]
But it will work. Maybe... For the time being, as a flight mode switch. And then we will wait for the remapping of channels ))

**Reply by ru13ab** [reply]
16ch mode - CH1-CH16 are sent 10-bit at the half rate

**Reply by ru13ab** [reply]
Apparently this is what we see in the picture above

**Reply by robhaswell** [reply]
Awesome yeah I guess 16CH 10-bit at half the rate would get me flying! Thanks!

**Reply by ru13ab** [reply]
Great news.  My transmitter hasn't arrived yet.  Can't try it myself.  Are you using MATEK CRSF-SBUS Converter and does it work?  Then this is the second great news.)

**Reply by robhaswell** [reply]
Unfortunately it seems like because the `IsArmed()` state is tied to CH5 still, the won't be suitable - as CH5 needs to be high for all the functions tied to it to work. I think it would fly but might not be safe.

**Reply by ru13ab** [reply]
Sorry.  My English is not that good.  I didn't read carefully.  It looks like you've already tried it. 
    I don't think it's possible to fly a helicopter so far that there will be problems with it.  Although other factors than distance can affect.  But I have not yet studied this issue carefully.

**Reply by robhaswell** [reply]
No your English is fine. I found that out since posting earlier.

Range - tbh my current DSMX setup is OK for LOS flight. But I want to move to ELRS for two reasons -

1. Homologate all models behind a single radio system
2. Attach FPV gear to my M2

For an FPV helicopter I would definitely want to know the ELRS link is operating as intended. Helis make a mess when they crash.

**Reply by ru13ab** [reply]
![Screenshot_20220703-175248_Discord](https://user-images.githubusercontent.com/85231786/177045250-a06bcfb0-9154-45b1-967d-5d6ae331a189.jpg)

**Reply by ru13ab** [reply]
You can try this, there is no other way

**Reply by robhaswell** [reply]
Thanks, I will try this.

**Comment by socalengineer** [comment]
I am using a Matek CRSF-PWM-C (10ch CRSF->PWM converter board) and I kept running into issues with using Ch5 because it is the ARM switch, and you cannot adjust Ch5 on the ELRS-R24-D (or ELRS-R24-S) to be mapped to a different channel.

I am using ELRS 2.5.1, and I wanted Channel 11 to be sent to Channel 5 (aux 1) so that I would get to use the full 10channels of the PWM output of the receiver.

OTA.cpp:
UnpackChannelDataHybridWide function line 246:
Comment out: crsf->PackedRCdataOut.ch4 = BIT_to_CRSF((switchByte & 0b10000000) >> 7);
Replace line 290: // crsf->PackedRCdataOut.ch10 = switchValue;
with new line: crsf->PackedRCdataOut.ch4 = switchValue;

And then I did the same thing in the UnpackChannelDataHybridSwitch8 function
Commment out: crsf->PackedRCdataOut.ch4 = BIT_to_CRSF((switchByte & 0b01000000) >> 6);
Replace line 214: crsf->PackedRCdataOut.ch10 = switchValue;
with new line: crsf->PackedRCdataOut.ch4 = switchValue;


Now channel 5 pwm output will be channel 11. Channel 11 doesn't exist on the PWM board, so no harm. Works great for me, now I have a fully functional 10ch receiver for my fixed wing. Before it felt like a crippled 9 channel receiver. Not sure if this is the right code to edit, but it works great!

**Reply by jurgelenas** [reply]
> I am using a Matek CRSF-PWM-C (10ch CRSF->PWM converter board) and I kept running into issues with using Ch5 because it is the ARM switch, and you cannot adjust Ch5 on the ELRS-R24-D (or ELRS-R24-S) to be mapped to a different channel.
> 
> I am using ELRS 2.5.1, and I wanted Channel 11 to be sent to Channel 5 (aux 1) so that I would get to use the full 10channels of the PWM output of the receiver.
> 
> OTA.cpp: UnpackChannelDataHybridWide function line 246: Comment out: crsf->PackedRCdataOut.ch4 = BIT_to_CRSF((switchByte & 0b10000000) >> 7); Replace line 290: // crsf->PackedRCdataOut.ch10 = switchValue; with new line: crsf->PackedRCdataOut.ch4 = switchValue;
> 
> And then I did the same thing in the UnpackChannelDataHybridSwitch8 function Commment out: crsf->PackedRCdataOut.ch4 = BIT_to_CRSF((switchByte & 0b01000000) >> 6); Replace line 214: crsf->PackedRCdataOut.ch10 = switchValue; with new line: crsf->PackedRCdataOut.ch4 = switchValue;
> 
> Now channel 5 pwm output will be channel 11. Channel 11 doesn't exist on the PWM board, so no harm. Works great for me, now I have a fully functional 10ch receiver for my fixed wing. Before it felt like a crippled 9 channel receiver. Not sure if this is the right code to edit, but it works great!

I think you can remap that using matek configurator: http://www.mateksys.com/?portfolio=crsf-pwm#tab-id-6

**Comment by socalengineer** [comment]
You’re right, you can do that through the configurator! This way you don’t
have to connect to the USB port and do that wiring though.
